### PR TITLE
WebXR Support

### LINKED ISSUE #1748: Add WebXR support
The engine currently supports WebVR. This is being [removed from Chrome in version 80](https://caniuse.com/#search=webvr). [WebXR](https://www.w3.org/TR/webxr/) is being [enabled (not behind a flag!) in Chrome 79](https://caniuse.com/#search=webxr).

Question:
Should the engine retain WebVR support for now and add WebXR support? The engine would then attempt to obtain access to the WebXR API first and fall back to WebVR if not present. Alternatively, WebVR support could be stripped out and WebXR added in its place. The former feels like the correct choice.

### PR DESCRIPTION
**Pull request by Maksims**
This PR removes VR APIs from engine, and adds XR support.
Currently it implements only starting/ending XR sessions (VR/AR). Input devices and reference space features will be added in another PR.

New APIs:
```javascript
// pc.XrManager as 'xr' on pc.Application
app.xr // access to XrManager
// properties
app.xr.supported // true if WebXR is supported
app.xr.active // true if WebXR Session is active
app.xr.type // null or string (pc.XRTYPE_*) of current active session
// methods
app.xr.isAvailable(type) // check if session of pc.XRTYPE_* is available on current platform
app.xr.start(camera, type, [callback]) // attempt to start XR session for CameraComponent and of pc.XRTYPE_*. With optional callback that can return Error if failed to start.
app.xr.end([callback]) // attempt to end XR session. With optional callback with Error if failed to end.
// events
app.xr.on('available', function (type, available) { } ) // event fired when availability of specified type is changed. Where type is one of pc.XRTYPE_* and available is a boolean.
app.xr.on('available:[type]', function (available) { } ) // event fired when availability of specified type is changed. Where [type] is one of pc.XRTYPE_* and available is a boolean.
app.xr.on('start', function () { } ) // event fired when session is started
app.xr.on('end', function () { } ) // event fired when session is ended

// pc.CameraComponent
entity.camera.startXr(type, [callback]) // start session from specified camera component, where type is one of pc.XRTYPE_*. With optional callback that can return Error if failed to start.
entity.camera.endXr([callback]) // attempt to end XR session. With optional callback with Error if failed to end.

// pc
pc.XRTYPE_INLINE // 'inline' - type of session which is always available
pc.XRTYPE_VR // 'immersive-vr' - type of session, offers best VR experience with access to most hardware sensors
pc.XRTYPE_AR // 'immersive-ar' - type of session, available on platforms with camera.
```

Fixes #1748

I confirm I have signed the [Contributor License Agreement](https://docs.google.com/a/playcanvas.com/forms/d/1Ih69zQfJG-QDLIEpHr6CsaAs6fPORNOVnMv5nuo0cjk/viewform).

### CODE REVIEW

**Comment by Maksims** [comment]
Some progress here, added ability to choose session type.
Fixed issues with switching between rendering targets (backbuffer of framebuffer for XR).
Fixed issues with canvas resizing.

Tested AR on Android with chrome://flags/#webxr enabled. Surprised to see how well it works already.
For testing, can use this project: https://playcanvas.com/project/660566/overview/webxr

**Comment by Maksims** [comment]
This PR already implements what WebVR was providing - starting/ending XR sessions.
It does not implements XR input devices, nor reference space features. This will be added in another PR.

So this can be merged, as it will provide ability to start experimenting with VR/AR.
It can be tested on this project: https://playcanvas.com/project/660566/overview/webxr

**Comment by willeastcott** [comment]
Great stuff!

The unit tests are failing. Can you fix please?

File: src/framework/application.js [hunk 1]
```diff
@@ -1565,29 +1577,6 @@ Object.assign(pc, function () {
...
-         * @name pc.Application#disableVr
-         * @description Destroy the {@link pc.VrManager}.
-         */
-        disableVr: function () {
-            if (this.vr) {
-                this.vr.destroy();
-                this.vr = null;
-            }
-        },
-
```
Comment (willeastcott) [inline src/framework/application.js]: For any API that you remove, can you add stub functions that print a debug warning to src\backwards-compatibility.js please?

Comment (Maksims) [reply src/framework/application.js]: Added some warnings.

**Comment by Maksims** [comment]
Added deprecation warnings, and fixed tests.

**Comment by willeastcott** [comment]
Excellent. Can you also update this please:

https://github.com/playcanvas/engine/blob/master/examples/graphics/virtual-reality.html

Maybe even consider creating a new subfolder 'xr' and move it there (we'll probably add an AR and other related examples too!). 😄

**Comment by willeastcott** [comment]
A few lint errors. You can auto-fix most/all:

`npm run lint -- --fix`

**Comment by Maksims** [comment]
Fixed lint and added two examples for basic VR and basic AR.
Also updated [Starter Kit: VR](https://playcanvas.com/project/435780/overview/starter-kit-vr) project in a **webxr** branch, so can be merged to **master** once engine is deployed with this PR.

File: src/framework/application.js [hunk 2]
```diff
@@ -1808,20 +1803,27 @@ Object.assign(pc, function () {
...
-            if (app.graphicsDevice.contextLost) {
+            if (app.graphicsDevice.contextLost)
                 return;
-            }
 
             // #ifdef PROFILER
             app._fillFrameStats(now, dt, ms);
             // #endif
 
+            if (frame) app.xr.calculateViews(frame);
```
Comment (willeastcott) [inline src/framework/application.js]: Why not put this inside the `if (frame)` condition below?

File: src/framework/components/camera/component.js [hunk 3]
```diff
@@ -134,45 +134,26 @@ Object.assign(pc, function () {
...
         }
     });
 
     /**
      * @readonly
-     * @name pc.CameraComponent#node
-     * @type {pc.GraphNode}
-     * @description Queries the camera's GraphNode. Can be used to get position and rotation.
+     * @name pc.CameraComponent#xr
+     * @type Boolean
```
Comment (willeastcott) [inline src/framework/components/camera/component.js]: Boolean -> boolean

File: src/framework/components/camera/component.js [hunk 4]
```diff
@@ -134,45 +134,26 @@ Object.assign(pc, function () {
...
+            return this.data.camera._node;
         }
     });
 
     /**
      * @readonly
-     * @name pc.CameraComponent#node
-     * @type {pc.GraphNode}
-     * @description Queries the camera's GraphNode. Can be used to get position and rotation.
+     * @name pc.CameraComponent#xr
```
Comment (willeastcott) [inline src/framework/components/camera/component.js]: Do you think `pc.CameraComponent#xr` will be confused with `pc.Application#xr`? In other words, might the developer assume it's another `pc.XrManager` instance? Alternative naming suggestions: `xrActive`, `xrEnabled`.

Comment (Maksims) [reply src/framework/components/camera/component.js]: Based on other naming conventions, do you think `isXr` is reasonable naming for this property?

Comment (willeastcott) [reply src/framework/components/camera/component.js]: Just chatted with the team. The most preferred naming is `pc.CameraComponent#xrEnabled`. Let's go with that. 😄

**Comment by Maksims** [comment]
Hm, noticed one issue with Box Projected CubeMaps, they are not projected correctly in VR.

**Comment by willeastcott** [comment]
Interesting. Screenshot?

**Comment by Maksims** [comment]
> Interesting. Screenshot?

box-project-wrong

**Comment by Maksims** [comment]
Fixed normals issue.

I believe this is ready now.

**Comment by willeastcott** [comment]
This is great - we're so close to merging. Last outstanding thing for me:

If I merged this and deployed as v1.25.0, what will VR developers face? Especially those with PlayCanvas projects based on the old WebVR integration?

* Do we need a migration guide for existing PlayCanvas VR apps?
* What about VR controllers? What has to be done to re-enable those?

**Comment by Maksims** [comment]
> If I merged this and deployed as v1.25.0, what will VR developers face? Especially those with PlayCanvas projects based on the old WebVR integration?

Currently this PR only enough for HMD (including mobile). WebVR is getting deprecated in most browsers, new Chrome wont even have its support behind the flag.

> * Do we need a migration guide for existing PlayCanvas VR apps?

We will need migration guide once we have all the functionality in place.

> * What about VR controllers? What has to be done to re-enable those?

Previously, WebVR had controllers implemented through GamePad API. WebXR implements own interfaces to interact with physical devices (even mouse and touch) and is done through XRInputSource.

**We can deprecate WebVR later, and deploy this, allowing developers to play with WebXR straight away. Later deprecate WebVR, as browsers slowly deprecate its APIs as well.**

**Comment by Maksims** [comment]
Updated: WebVR is brought back. So we will deprecate it later.
And slightly simplified API.

File: src/framework/components/camera/component.js [hunk 5]
```diff
@@ -556,6 +568,51 @@ Object.assign(pc, function () {
...
+         * @param {string} type - The type of session. Can be one of the following:
+         *
+         * * {@link pc.XR_TYPE_INLINE}: Inline - always available type of session. It has limited features availability and is rendered into HTML element.
+         * * {@link pc.XR_TYPE_IMMERSIVE_VR}: Immersive VR - session that provides exclusive access to VR device with best available tracking features.
+         * * {@link pc.XR_TYPE_IMMERSIVE_AR}: Immersive AR - session that provides exclusive access to VR/AR device that is intended to be blended with real-world environment.
+         *
+         * @param {pc.callbacks.XrError} [callback] - Optional callback function called once session is started. The callback has one argument Error - it is null if successfully started XR session.
+         * @example
+         * // On an entity with a camera component
+         * this.entity.camera.startXr(PC.XR_TYPE_IMMERSIVE_VR, function (err) {
```
Comment (willeastcott) [inline src/framework/components/camera/component.js]: PC -> pc

File: src/framework/components/camera/component.js [hunk 6]
```diff
@@ -556,6 +568,51 @@ Object.assign(pc, function () {
...
+        },
+
+        /**
+         * @function
+         * @name pc.CameraComponent#startXr
+         * @description Attempt to start XR session with this camera
+         * @param {string} type - The type of session. Can be one of the following:
+         *
+         * * {@link pc.XR_TYPE_INLINE}: Inline - always available type of session. It has limited features availability and is rendered into HTML element.
+         * * {@link pc.XR_TYPE_IMMERSIVE_VR}: Immersive VR - session that provides exclusive access to VR device with best available tracking features.
```
Comment (willeastcott) [inline src/framework/components/camera/component.js]: In the interests of a simpler API, what about: `pc.XR_TYPE_IMMERSIVE_VR` -> `pc.XRTYPE_VR`? Same for AR. (BTW, the general naming convention is `pc.THISISTHETYPE_THISISTHESUBTYPE`, so one underscore. Yes, I know some enums break this).

**Comment by willeastcott** [comment]
I definitely like how you've simplified some of the API new - nice move.

File: examples/xr/augmented-reality-basic.html [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+            var SIZE = 4;
+            for (var x = 0; x < SIZE; x++) {
+                for (var y = 0; y < SIZE; y++) {
+                    createCube(2*x - SIZE, -1, 2*y - SIZE);
+                }
+            }
+
+            if (app.xr.supported) {
+                var activate = function () {
+                    if (app.xr.isAvailable(pc.XR_TYPE_IMMERSIVE_AR)) {
```
Comment (willeastcott) [inline examples/xr/augmented-reality-basic.html]: So you have `app.xr.supported` and then `app.xr.isAvailable(pc.XR_TYPE_IMMERSIVE_AR)`. For consistency, and better readability, what about `app.xr.supports(pc.XRTYPE_AR)`

Comment (Maksims) [reply examples/xr/augmented-reality-basic.html]: Support is static property, that is set before page is loaded. Which defines if platform supports WebXR API at all.
And "availability" - is what type of session currently is available, which can change during page life based on what devices are plugged in, etc.
I've split them intentionally, so not to confuse support with availability.

Support - cannot change, only browser upgrade can.
Availability - can change dynamically during page life.

File: src/framework/components/camera/component.js (hunk 4 above)
Comment (Maksims) [reply src/framework/components/camera/component.js]: I actually couldn't think of use case for this property, so removed it entirely.

**Comment by Maksims** [comment]
Tested on Oculus Quest, works as intended.

**Pull request merged**